### PR TITLE
Define Schema using decorators

### PR DESCRIPTION
**Pull request by marcus-pousette**
Why? 
- Ergonomics could be improved (I do not want to write duplicate code)
- Handling enums is confusing

Changes: 
- Added decorators so we can define a mapping from classes to schemas without leaving the class definintion. 
- **Breaking**. The 'borshSerialize' function is changed to include the schema map. 
- Refactoring. Moving out writing and reading functions. Moving BorshError to separate file
- Adding support for defining tests in typescript

Todos (maybe): 
- Deserialization has not yet been considered. 

See tests for all examples, but just a few examples: 

**Enum, variant at instruction "slot" 1.** 

```typescript
import { generateSchemas, StructKind, Field, Variant } from "../schema";

@Variant(1)
class TestEnum {
    @Field({ type: 'u8' })
    public a: number;

    constructor(a: number) {
        this.a = a

    }
}

class TestStruct {
    @Field({ type: TestEnum })
    public enum: TestEnum;

    constructor(value: TestEnum) {
        this.enum = value
    }
}
const instance = new TestStruct(new TestEnum(4));
const generatedSchemas = generatesSchema([TestStruct])
const buf = serialize(generatedSchemas, instance);
expect(buf).toEqual(Buffer.from([1, 4]));

```

**Nested Schema generation for structs**

```typescript
import { generateSchemas, StructKind, Field } from "../schema";

class InnerStruct {
    @Field({ type: 'typeB' })
    public b: number;

}

class TestStruct {
    @Field({ type: InnerStruct })
    public a: InnerStruct;

}

const generatedSchemas = generateSchemas([TestStruct])
expect(generatedSchemas.get(TestStruct)).toEqual({
    kind: 'struct',
    fields: [
        ['a', InnerStruct],
    ]
});
expect(generatedSchemas.get(InnerStruct)).toEqual({
    kind: 'struct',
    fields: [
        ['b', 'typeB'],
    ]
});
```

**Option**

```typescript
import { generateSchemas, StructKind, Field } from "../schema";

class TestStruct {
  @Field({ type: 'u8', option: true })
  public a: number;

}
const schema: StructKind = generateSchemas([TestStruct]).get(TestStruct)
expect(schema).toEqual({
  fields: [
      [
          "a",
          {
              kind: 'option',
              type: 'u8'
          },
      ]
  ],
  kind: "struct",
});
```
**Explicit serialization order of fields**

```typescript
import { generateSchemas, StructKind, Field } from "../schema";

class TestStruct {
    @Field({ type: 'u8', index: 1 })
    public a: number;


    @Field({ type: 'u8', index: 0 })
    public b: number;
}
const schema: StructKind = generateSchemas([TestStruct]).get(TestStruct)
expect(schema).toEqual({
    fields: [
        [
            "b",
            "u8",
        ],
        [
            "a",
            "u8",
        ],
    ],
    kind: "struct",
});
```

Please let me know (quite quickly) if you want these kinds of ergonomics in the library. If so, and you are actively working on this repository, I will put some more time this week into this PR.

### CODE REVIEW

**Comment by dgabriele** [comment]
Any update on this? Would be great to serialize enums like this.

File: lib/schema.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+const STRUCT_META_DATA_SYMBOL = '__borsh_struct_metadata__';
+const structMetaDataKey = (constructorName) => {
+    return STRUCT_META_DATA_SYMBOL + constructorName;
+};
+/**
+ *
+ * @param kind 'struct' or 'variant. 'variant' equivalnt to Rust Enum
+ * @returns Schema decorator function for classes
+ */
+exports.Variant = (index) => {
```
Comment (dgabriele) [inline lib/schema.js:16]: Can we use lowercase property names? `@variant` and `@field` instead of class-case?

File: CODEOWNERS [hunk 2]
```diff
@@ -1 +1 @@
-* @volovyk-s @ailisp
+* @marcus-pousette
```
Comment (volovyk-s) [inline CODEOWNERS:1]: You should be a part of NEAR Inc to become a maintainer ;) Ping me if you are interested.

**Review by volovyk-s — COMMENTED** [review]
Hey @marcus-pousette ! Thank you for your contribution. This library is actively used, but not in active development. Anyway, this change looks awesome and I will play with it. Have you added all the things that were planned, is it a final version?

File: CODEOWNERS (hunk 2 above)
Comment (marcus-pousette) [reply CODEOWNERS:1]: Hey thanks for the response! I see. As I needed the changes quite urgently I started to maintain my own fork [https://github.com/the-solvei/borsh-ts](https://github.com/the-solvei/borsh-ts). I fixed a few bugs and also made additional improvements (which means this PR is outdated). I would have to rebase those changes to this branch. If you would want to play around with a working solution I would recommend you to try out my forked version. 

I will give you a ping, regarding maintainer role 👍

Comment (volovyk-s) [reply CODEOWNERS:1]: Awesome! I will take a look.
Ping me after you rebase it.

Comment (AlaaZorkane) [reply CODEOWNERS:1]: Hey @marcus-pousette, any plans to publish to npm?

Comment (marcus-pousette) [reply CODEOWNERS:1]: @AlaaZorkane. Sorry for late reply. Yes! [it is live](https://github.com/dao-xyz/borsh-ts)

**Comment by ailisp** [comment]
I think our decision is to have a JS compatible borsh, so TS decorators won't fit our goal. But that's not to say it's not a good approach. For new comers who use TS, @marcus-pousette implemented https://github.com/dao-xyz/borsh-ts is a great library that you can consider. If you also need to use borsh in JS, then near/borsh-js is for you.

**Comment by gagdiez** [comment]
closing this PR since a different approach was maintained, and anyway the use of enums has been simplified.

**Pull request closed without merging**